docs/en/04_Changelogs/4.0.0.md

```diff
@@ -92,6 +93,138 @@ For a full list of renamed classes, check the `.upgrade.yml` definitions in each
 
 The rename won't affect class-based permission codes or database table names.
 
+##### Use ::class whenever possible
+
+Prefer:
+ ```php
+ private static $has_one = [
+   'MyObject' => MyObject::class
+ ];
+ ```
+ to
+ 
+ ```php
+private static $has_one = [
+  'MyObject' => 'My\Project\MyObject'
+];
+```
+
+##### Use Injector wherever you can
+
+This can help with reducing changes from 3.x to 4.x by allowing your code to 
+continue to use `MyClassName`, while you tell the injector what the FQ class name is:
+
+```yaml
+SilverStripe\Core\Injector\Injector:
+  MyClassName:
+    class: Me\MyModule\MyClassName # fully qualified!
+```
+
+##### Namespaced table names
+
+Keep in mind that table names are namespaced too. To ease data migrations, use the `$table_name` config property of your `DataObject` subclasses.
+This allows you to customise the table name to something simpler, or perhaps identical to its 3.x name.
+```php
+namespace SilverStripe\BannerManager;
+
+use SilverStripe\ORM\DataObject;
+
+class BannerImage extends DataObject
+{
+	private static $table_name = 'BannerImage';
+}
+```
+
+In order to ensure you are using the correct table for any class a new [DataObjectSchema](api:SilverStripe\ORM\DataObjectSchema) service
+is available to manage these mappings (see [versioned documentation](/developer_guides/model/data_model_and_orm)).
+
+
+```php
+public function countDuplicates($model, $fieldToCheck)
+{
+	$table = SilverStripe\ORM\DataObject::getSchema()->tableForField($model, $field);
+	$query = new SilverStripe\ORM\Queries\SQLSelect();
+	$query->setFrom("\"{$table}\"");
+	$query->setWhere(["\"{$table}\".\"{$field}\"" => $model->$fieldToCheck]);
+	return $query->count();
+}
+```
+
+##### Class name remapping
+
+If you've namespaced one of your custom page types, you may notice a message in the CMS 
+ telling you it's obsolete. This is likely because the `ClassName` 
+field in the `SiteTree` table still contains the singular model name, e.g. `BlogPost`
+and that when you change it to `SilverStripe\Blog\Model\BlogPost` then everything 
+works again.
+
+Luckily the `dev/build` task is configured to look for a legacy class name mapping 
+configuration setting and will update this for you automatically. For an example 
+take a look at `_config/legacy.yml` in the CMS module. The Blog module, for instance, uses 
+this:
+
+```yaml
+SilverStripe\ORM\DatabaseAdmin:
+  classname_value_remapping:
+    Blog: SilverStripe\Blog\Model\Blog
+    BlogCategory: SilverStripe\Blog\Model\BlogCategory
+    BlogPost: SilverStripe\Blog\Model\BlogPost
+    BlogTag: SilverStripe\Blog\Model\BlogTag
+```
+
+##### Other namespacing gotchas
+
+You'll need to update any strings that represent class names and make sure they're fully
+qualified. In particular, relationship definitions such as `has_one` and `has_many` will need
+to be updated to refer to fully qualified class names.
+
+```ss
+<!-- before -->
+private static $has_one = [
+  'MyRelation' => 'MyObject',
```

Comment:
Upgrader does a lot of this automatically. I strongly suggest to put this into a "how to use upgrader to automatically upgrade your code" block.